web-common/src/features/alerts/alert-preview-data.ts

```diff
@@ -0,0 +1,135 @@
+import { sanitiseExpression } from "@rilldata/web-common/features/dashboards/stores/filter-utils";
+import type { MetricsExplorerEntity } from "@rilldata/web-common/features/dashboards/stores/metrics-explorer-entity";
+import type { TimeControlState } from "@rilldata/web-common/features/dashboards/time-controls/time-control-store";
+import { getOffset } from "@rilldata/web-common/lib/time/transforms";
+import { TimeOffsetType } from "@rilldata/web-common/lib/time/types";
+import {
+  createQueryServiceMetricsViewAggregation,
+  queryServiceMetricsViewAggregation,
+  type V1Expression,
+  type V1MetricsViewAggregationRequest,
+  type V1MetricsViewAggregationResponseDataItem,
+} from "@rilldata/web-common/runtime-client";
+import type {
+  CreateQueryOptions,
+  CreateQueryResult,
+} from "@tanstack/svelte-query";
+import { derived } from "svelte/store";
+
+export type AlertPreviewParams = {
+  measure: string;
+  dimension: string;
+  criteria: V1Expression | undefined;
+  splitByTimeGrain: string | undefined;
+};
+export type AlertPreviewResponse = {
+  rows: V1MetricsViewAggregationResponseDataItem[];
+  schema: VirtualizedTableColumns[];
+};
+
+export function getAlertPreviewData(
+  ctx: StateManagers,
+  params: AlertPreviewParams,
+): CreateQueryResult<AlertPreviewResponse> {
+  return derived(
+    [
+      ctx.runtime,
+      ctx.metricsViewName,
+      ctx.dashboardStore,
+      ctx.selectors.timeRangeSelectors.timeControlsState,
+      measureFilterResolutionsStore(ctx),
+    ],
+    (
+      [
+        runtime,
+        metricsViewName,
+        dashboard,
+        timeControls,
+        resolvedMeasureFilters,
+      ],
+      set,
+    ) => {
+      return createQueryServiceMetricsViewAggregation(
+        runtime.instanceId,
+        metricsViewName,
+        getAlertPreviewQueryRequest(
+          params,
+          dashboard,
+          timeControls,
+          resolvedMeasureFilters,
+        ),
+        {
+          query: getAlertPreviewQueryOptions(
+            ctx,
+            params,
+            timeControls,
+            resolvedMeasureFilters,
+          ),
+        },
+      ).subscribe(set);
+    },
+  );
+}
+
+function getAlertPreviewQueryRequest(
+  params: AlertPreviewParams,
+  dashboard: MetricsExplorerEntity,
+  timeControls: TimeControlState,
+  resolvedMeasureFilters: ResolvedMeasureFilter,
+): V1MetricsViewAggregationRequest {
+  return {
+    measures: [{ name: params.measure }],
+    dimensions: params.dimension ? [{ name: params.dimension }] : [],
+    where: sanitiseExpression(
+      dashboard.whereFilter,
+      resolvedMeasureFilters.filter,
+    ),
+    having: sanitiseExpression(undefined, params.criteria),
+    timeRange: {
+      start:
+        params.splitByTimeGrain && timeControls.timeEnd
+          ? getOffset(
+              new Date(timeControls.timeEnd),
+              params.splitByTimeGrain,
+              TimeOffsetType.SUBTRACT,
+            )
+          : timeControls.timeStart,
```

Comment:
Is this offset actually needed? Adding the time + time grain to the `dimensions` would make this unnecessary, right?

Reply:
No was getting some error with pivot query. So had this for just the last time segment